compile error:  std::result_of - deprecated in c++17 and removed in c++20

The use of std::result_of in the project prevents it from being built in C++20 projects;

e.g

iceoryx_hoofs/cxx/type_traits.hpp(66,56): error C2039: 'result_of': is not a member of 'std'


Would you consider replacing it with std::invoke_result or suitable wrappers ?

In the meantime for vc one can reintroduce the templates and get it to compile by defining _HAS_DEPRECATED_RESULT_OF

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -43,6 +43,7 @@
 - Catch deserialization errors for enums in publisher and subscriber options [\#989](https://github.com/eclipse-iceoryx/iceoryx/issues/989)
 - Fix linker error on QNX [\#1013](https://github.com/eclipse-iceoryx/iceoryx/issues/1013)
 - When posix mutex fails a correct error message is reported on the console [\#999](https://github.com/eclipse-iceoryx/iceoryx/issues/999)
+- Only use `std::result_of` for C++14 to be able to use iceoryx in C++20 projects [\#1076](https://github.com/eclipse-iceoryx/iceoryx/issues/1076)
 
 **Refactoring:**
 

--- a/iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp
+++ b/iceoryx_hoofs/include/iceoryx_hoofs/cxx/type_traits.hpp
@@ -1,5 +1,5 @@
 // Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
-// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+// Copyright (c) 2021 - 2022 by Apex.AI Inc. All rights reserved.
 //
 // Licensed under the Apache License, Version 2.0 (the "License");
 // you may not use this file except in compliance with the License.
@@ -63,7 +63,12 @@ struct is_invocable
     // This variant is chosen when Callable(ArgTypes) successfully resolves to a valid type, i.e. is invocable.
     /// @note result_of is deprecated, switch to invoke_result in C++17
     template <typename C, typename... As>
+#if __cplusplus < 201703L
     static constexpr std::true_type test(typename std::result_of<C(As...)>::type*) noexcept
+    // use std::invoke_result so that iceoryx can be used in c++20 projects
+#elif __cplusplus >= 201703L
+    static constexpr std::true_type test(typename std::invoke_result<C, As...>::type*) noexcept
+#endif
     {
         return {};
     }
@@ -90,7 +95,13 @@ struct is_invocable_r
 {
     template <typename C, typename... As>
     static constexpr std::true_type
+#if __cplusplus < 201703L
     test(std::enable_if_t<std::is_convertible<typename std::result_of<C(As...)>::type, ReturnType>::value>*) noexcept
+    // use std::invoke_result so that iceoryx can be used in c++20 projects
+#elif __cplusplus >= 201703L
+    test(
+        std::enable_if_t<std::is_convertible<typename std::invoke_result<C, As...>::type, ReturnType>::value>*) noexcept
+#endif
     {
         return {};
     }

